### DISCUSSION TITLE
Test core count and hyperthreadnig

### DISCUSSION THREAD

**Discussion by austindonisan** [gunnarmorling/1brc#443] [Rule clarifications] [answered]
It says the program is tested using 8 cores of an EPYC 7502P. How is that implemented?

Does Runtime.getRuntime().availableProcessors() return 8? Or 16 since hyperthreading is enabled?

Edit: availableProcessors() will definitely always return 8. It's just question if those are all physical cores or not.

@gunnarmorling Can you confirm if hyperthreading is enabled on the test server?

My implementation is efficient enough at exploiting ILP that running extra threads with hyperthreading just slows it down due to cache contention.

Edit: It seems like server CPUS all have their physical cores listed first, for both AMD and Intel, even for multiple-socket machines.

**Comment by lehuyduc** [comment, marked as answer]
```
    # Linux platform
    # prepend this with numactl --physcpubind=0-7 for running it only with 8 cores
    numactl --physcpubind=0-7 hyperfine $HYPERFINE_OPTS "$TIMEOUT ./calculate_average_$fork.sh 2>&1"
```
You can check in `evaluate.sh`

**Comment by austindonisan** [comment]
So if hyperthreading is enabled it will only run on 4 cores with 8 threads.

However the script also tries to print a warning if hyperthreading is enabled, so it's seems like maybe it's not enabled on the test machine.

**Reply by lehuyduc** [reply]
Actually no. Virtual core is from N -> 2*N-1. Physical core 0 is represented by 2 virtual cores (0,16), physical core 1 by (1,17), ...

So `0-7` means it can run on 8 different physical cores. But `numactl` causes lower performance than running 8 threads manually in your code, i'm not sure why yet.

**Reply by austindonisan** [reply]
Every consumer Intel CPU I've ever had has them numbered 0-1, 2-3, 4-5 etc. AMD or server CPUs could be different, though.

**Reply by lehuyduc** [reply]
Oh, what I said applies to AMD
